### PR TITLE
Add JWT issuing and validating support

### PR DESCRIPTION
**Pull request by isharak**
## Purpose
This module will support for issuing and validating JWT tokens.

### CODE REVIEW

File: stdlib/ballerina-builtin/src/main/ballerina/ballerina/jwt/issuer.bal [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    string keyPassword;
+}
+
+@Description {value:"Issue a JWT token"}
+@Param {value:"header: Header object"}
+@Param {value:"payload: Payload object"}
+@Param {value:"config: JWTIssuerConfig object"}
+@Return {value:"string: JWT token string"}
+@Return {value:"error: If token validation fails "}
+public function issue (Header header, Payload payload, JWTIssuerConfig config) (string, error) {
```
Comment (mefarazath) [inline stdlib/ballerina-builtin/src/main/ballerina/ballerina/jwt/issuer.bal]: Shouldn't  we add the jti value here as well?

File: stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/jwt/crypto/JWSSigner.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.ballerinalang.nativeimpl.jwt.crypto;
+
+/**
+ * JSON web signature
```
Comment (nadeeshaan) [inline stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/jwt/crypto/JWSSigner.java]: Missing period in doc comment

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/jwt/JwtTest.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        Path sourceRoot = Paths.get(resourceRoot, "test-src", "jwt");
+        compileResult = BCompileUtil.compile(sourceRoot.resolve("jwt-test.bal").toString());
+    }
+
+    @Test(priority = 1, description = "Test case for issuing JWT token with valid data")
+    public void testIssueJwt() throws Exception {
+        BValue[] returns = BRunUtil.invoke(compileResult, "testIssueJwt");
+        Assert.assertTrue(returns[0] instanceof BString);
+        Assert.assertNull(returns[1]);
+        jwtToken = returns[0].stringValue();
```
Comment (djkevincr) [inline tests/ballerina-test/src/test/java/org/ballerinalang/test/jwt/JwtTest.java]: Here we set value for jwtToken but it doesn't have any further usages. Am I missing something here?

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/jwt/crypto/JWTVerifierTest.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import org.testng.annotations.Test;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.InputStream;
+import java.security.KeyStore;
+import java.security.cert.Certificate;
+import java.security.interfaces.RSAPublicKey;
+
+public class JWTVerifierTest {
```
Comment (djkevincr) [inline tests/ballerina-test/src/test/java/org/ballerinalang/test/jwt/crypto/JWTVerifierTest.java]: Shall we add class level java doc?

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/jwt/crypto/JWTSignerTest.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.InputStream;
+import java.security.KeyStore;
+import java.security.PrivateKey;
+
+public class JWTSignerTest {
```
Comment (djkevincr) [inline tests/ballerina-test/src/test/java/org/ballerinalang/test/jwt/crypto/JWTSignerTest.java]: Shall we add class level java doc?

File: stdlib/ballerina-builtin/src/main/ballerina/ballerina/jwt/validator.bal [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+        } else if (key == NBF) {
+            var value = jwtPayloadJson[key].toString();
+            var intVal, _ = <int>value;
+            jwtPayload.nbf = intVal;
+        } else if (key == IAT) {
+            var value = jwtPayloadJson[key].toString();
+            var intVal, _ = <int>value;
+            jwtPayload.iat = intVal;
+        }
+        else {
```
Comment (djkevincr) [inline stdlib/ballerina-builtin/src/main/ballerina/ballerina/jwt/validator.bal]: Shall we fix formatting here?

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/jwt/crypto/JWTVerifierTest.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+    private RSAPublicKey getRSAPublicKey() throws Exception {
+        KeyStore trustStore;
+        InputStream file = new FileInputStream(new File(getClass().getClassLoader().getResource(
+                "datafiles/security/keyStore/ballerinaTruststore.p12").getPath()));
+        trustStore = java.security.KeyStore.getInstance("pkcs12");
+        trustStore.load(file, "ballerina".toCharArray());
+        Certificate publicCertificate = trustStore.getCertificate("ballerina");
+        return (RSAPublicKey) publicCertificate.getPublicKey();
+
```
Comment (djkevincr) [inline tests/ballerina-test/src/test/java/org/ballerinalang/test/jwt/crypto/JWTVerifierTest.java]: Additional line?

File: stdlib/ballerina-builtin/src/main/ballerina/ballerina/jwt/issuer.bal (hunk 1 above)
Comment (isharak) [reply stdlib/ballerina-builtin/src/main/ballerina/ballerina/jwt/issuer.bal]: You can set that within payload

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/jwt/JwtTest.java (hunk 3 above)
Comment (isharak) [reply tests/ballerina-test/src/test/java/org/ballerinalang/test/jwt/JwtTest.java]: this is the JWT token that used to test verify method.

File: stdlib/ballerina-builtin/src/main/ballerina/ballerina/jwt/issuer.bal (hunk 1 above)
Comment (mefarazath) [reply stdlib/ballerina-builtin/src/main/ballerina/ballerina/jwt/issuer.bal]: :+1:

File: stdlib/ballerina-builtin/src/main/ballerina/ballerina/jwt/issuer.bal [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+function createHeader (Header header) (string) {
+    json headerJson = {};
+    headerJson[ALG] = header.alg;
+    headerJson[TYP] = "JWT";
+    headerJson = addMapToJson(headerJson, header.customClaims);
+    return urlEncode(util:base64Encode(headerJson.toString()));
+}
+
+function createPayload (Payload payload) (string) {
+    json payloadJson = {};
```
Comment (mefarazath) [inline stdlib/ballerina-builtin/src/main/ballerina/ballerina/jwt/issuer.bal]: As an improvement let's validate the payload for mandatory attributes.

**Pull request merged**